db/kv_checksum.h

```diff
@@ -0,0 +1,307 @@
+//  Copyright (c) 2020-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// This file contains classes containing fields to protect individual entries.
```

Comment:
@pdillinger - I got things working based on your suggested class design. Please take a look and LMK if it's what you had in mind/any further suggestions.

Again, I assumed the protection bytes per entry overhead is not hugely important at this stage (`WriteBatch`). In the latest update, only eight bytes per key is supported. There is some work in this PR towards supporting smaller powers of two, but I hope to defer finishing it to later.

Reply:
Do you mean to pass around the expected numerical value directly, or just replace `ProtectionInfo::buf_` with a numerical value? I think the `ProtectionInfo*` wrapper classes are useful for embedding the coverage info in the type.

If we do keep the `ProtectionInfo*` classes, I am not sure about abandoning `reinterpret_cast`. My first stab at this design included a class for each step of adding/removing protection info (i.e., `ProtectionInfo`, `ProtectionInfoK`, ..., `ProtectionInfoKVOTC`). Then, it was kind of convenient to avoid accessing the buffer as `kvot_.kvo_.kv_.k_.buf_`. It was and still is also useful for invoking the default copy constructor of a more-protected type (see any of the `Protect*()` implementations).